mockgcp/mockapikeys/apikeysv2.go

```diff
@@ -0,0 +1,191 @@
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	longrunning "google.golang.org/genproto/googleapis/longrunning"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+	"google.golang.org/protobuf/proto"
+	"google.golang.org/protobuf/types/known/fieldmaskpb"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+
+	pb "github.com/GoogleCloudPlatform/k8s-config-connector/mockgcp/generated/mockgcp/api/apikeys/v2"
+)
+
+type APIKeysV2 struct {
+	*MockService
+	pb.UnimplementedApiKeysServer
+}
+
+func (s *APIKeysV2) GetKey(ctx context.Context, req *pb.GetKeyRequest) (*pb.Key, error) {
+	name, err := s.parseAPIKeyName(req.Name)
+	if err != nil {
+		return nil, err
+	}
+
+	fqn := name.String()
+
+	obj := &pb.Key{}
+	if err := s.storage.Get(ctx, fqn, obj); err != nil {
+		if apierrors.IsNotFound(err) {
+			return nil, status.Errorf(codes.NotFound, "key %q not found", name)
+		} else {
+			return nil, status.Errorf(codes.Internal, "error reading key: %v", err)
+		}
+	}
+
+	return obj, nil
+}
+
+func (s *APIKeysV2) GetKeyString(ctx context.Context, req *pb.GetKeyStringRequest) (*pb.GetKeyStringResponse, error) {
+	name, err := s.parseAPIKeyName(req.Name)
+	if err != nil {
+		return nil, err
+	}
+
+	fqn := name.String()
+
+	obj := &pb.Key{}
+	if err := s.storage.Get(ctx, fqn, obj); err != nil {
+		if apierrors.IsNotFound(err) {
+			return nil, status.Errorf(codes.NotFound, "key %q not found", name)
+		} else {
+			return nil, status.Errorf(codes.Internal, "error reading key: %v", err)
+		}
+	}
+
+	// TODO: Do we need a real value here?
+	keyString := "dummy-encrypted-value"
+	return &pb.GetKeyStringResponse{
+		KeyString: keyString,
+	}, nil
+}
+
+func (s *APIKeysV2) CreateKey(ctx context.Context, req *pb.CreateKeyRequest) (*longrunning.Operation, error) {
+	keyID := req.KeyId
+	if keyID == "" {
+		keyID = fmt.Sprintf("key-%d", time.Now().UnixNano())
+	}
+	reqName := req.Parent + "/keys/" + keyID
+	name, err := s.parseAPIKeyName(reqName)
+	if err != nil {
+		return nil, err
+	}
+
+	fqn := name.String()
+
+	obj := proto.Clone(req.Key).(*pb.Key)
+	obj.Name = fqn
+
+	if err := s.storage.Create(ctx, fqn, obj); err != nil {
+		return nil, status.Errorf(codes.Internal, "error creating key: %v", err)
+	}
+
+	return s.operations.NewLRO(ctx)
+}
+
+func (s *APIKeysV2) UpdateKey(ctx context.Context, req *pb.UpdateKeyRequest) (*longrunning.Operation, error) {
+	keyName := req.GetKey().GetName()
+
+	name, err := s.parseAPIKeyName(keyName)
+	if err != nil {
+		return nil, err
+	}
+
+	fqn := name.String()
+
+	obj := &pb.Key{}
+	if err := s.storage.Get(ctx, fqn, obj); err != nil {
+		if apierrors.IsNotFound(err) {
+			return nil, status.Errorf(codes.NotFound, "key %q not found", name)
+		} else {
+			return nil, status.Errorf(codes.Internal, "error reading key: %v", err)
+		}
+	}
+
+	// From the proto:
+	// You can update only the `display_name`, `restrictions`, and `annotations` fields.
+
+	// See docs for UpdateMask
+	updateMask := req.GetUpdateMask()
```

Comment:
Again just for self-education purpose: How do we know that an update mask is needed if we happen to update the right field in the mock test (i.e. didn't trigger any test failure)?

Reply:
This is just a first-pass implementation based on what the docs on the proto describe.  There was actually some fun here about whether it is `display_name` or `displayName`, which I hope we'll be able to validate with the work in #957

Reply:
Right now it is "make the samples/fixtures tests pass".

Once we get #957, I think we'll be able to define a higher bar: capture the HTTP from real-gcp, make the mockgcp produce the same HTTP output (and make the samples/fixtures pass).  That is trial & error, but I believe we now call that "test driven development" ;-)

Yes, this requires the contributor to understand the API, but only the methods we are actually using.